kernel/src/path.rs

```diff
@@ -1,136 +1,178 @@
-    let path = location.path();
-    get_filename(path)
-        .and_then(|f| f.split_once('.'))
-        .and_then(|(name, _)| get_version_opt(Some(name), VERSION_LEN))
+impl AsUrl for FileMeta {
+    fn as_url(&self) -> &Url {
+        &self.location
+    }
 }
 
-impl<'a> LogPath<'a> {
-    pub(crate) fn new(url: &'a Url) -> Self {
-        let filename = get_filename(url.path());
-        let version_str = filename.and_then(|f| f.split_once('.'));
-        let version = version_str.and_then(|(name, _)| get_version_opt(Some(name), VERSION_LEN));
-
-        let mut is_commit = false;
-        let mut is_checkpoint = false;
-        let mut compacted_to_version = None;
-        if version.is_some() {
-            // could be a checkpoint or commit file, let's check
-            let (_, suffix) = version_str.unwrap(); // safe, version.is_some()
-            is_commit = suffix == "json"; // if we were just [version].json, we're a commit file
-
-            if !is_commit && suffix.starts_with("checkpoint.") {
-                // check if name is just [version].checkpoint.parquet, i.e. we have a classic checkpoint
-                let rest = &suffix[11..]; // strip off the "checkpoint." which is 11 chars
-                is_checkpoint = rest == "parquet";
-                if !is_checkpoint {
-                    // test if we're a multipart checkpoint
-                    let mut split = rest.splitn(3, '.');
-                    let (checkpoint_index, checkpoint_max, ext) = (
-                        get_version_opt(split.next(), MULTIPART_PART_LEN),
-                        get_version_opt(split.next(), MULTIPART_PART_LEN),
-                        split.next(),
-                    );
-                    is_checkpoint = checkpoint_index.is_some()
-                        && checkpoint_max.is_some()
-                        && ext == Some("parquet");
-                }
+impl<Location: AsUrl> ParsedLogPath<Location> {
+    // NOTE: We can't actually impl TryFrom because Option<T> is a foreign struct even if T is local.
+    pub fn try_from(location: Location) -> DeltaResult<Option<ParsedLogPath<Location>>> {
+        let url = location.as_url();
+        let filename = url
+            .path_segments()
+            .ok_or_else(|| Error::invalid_log_path(url))?
+            .last()
+            .unwrap() // "the iterator always contains at least one string (which may be empty)"
+            .to_string();
+        if filename.is_empty() {
+            return Err(Error::invalid_log_path(url));
+        }
+
+        let mut split = filename.split('.');
+
+        // NOTE: str::split always returns at least one item, even for the empty string.
+        let version = split.next().unwrap();
+
+        // Every valid log path starts with a numeric version part. If version parsing fails, it
+        // must not be a log path and we simply return None. However, it is an error if version
+        // parsing succeeds for a wrong-length numeric string.
```

Comment:
are the semantics here that `None` signifies it is a valid path but just not a log path? And err cases are actual parsing errors?

Reply:
Exactly. The unit tests cover a variety of both situations.